### PR TITLE
Potus: Fixes #1863

### LINKED ISSUE #1863: POTUS: Should not trigger for "vice president of the united states"
@numbertheory 
Searching for "vice president of the united states" triggers this instant answer, giving the incorrect output of Barack Obama

------
IA Page:  http://duck.co/ia/view/potus

### PR DESCRIPTION
**Pull request by hemanth**
^
Fixes #1863 

-----
IA Page: https://duck.co/ia/view/potus

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata post coming soon

**Comment by moollaza** [comment]
@hemanth please make sure to put "Fixes #1863 " in your PR description so the issue will close automatically when this is merged

**Comment by hemanth** [comment]
@moollaza Yup, but this is more like a quick review wrt [this](https://github.com/duckduckgo/zeroclickinfo-goodies/issues/1863#issuecomment-165329558)

File: lib/DDG/Goodie/POTUS.pm [hunk 1]
```diff
@@ -27,17 +27,12 @@ my $prez_count = scalar @presidents;
...
-    //gix;
+    return unless my ($tense, $index, $text) = $rem  =~ /^(?:who (?:is|was) the )?(\d+(?:st|nd|rd|th))?(?:potus|president of the (?:us|united states|usa)) ?(\d+)?$/;
 
     my $num = ($rem =~ /^\d+$/) ? $rem : words2nums($rem) || $prez_count;
-    my $index = $num - 1;
-    return if $index < 0 or $index > $#presidents;
-
+#   my $index = $num - 1;
+#   return if $index < 0 or $index > $#presidents;
+    print $index;
```
Comment (moollaza) [inline lib/DDG/Goodie/POTUS.pm]: You probably want `warn($index);`

Comment (hemanth) [inline lib/DDG/Goodie/POTUS.pm]: Yup, realised :angel:

**Comment by jagtalon** [comment]
@hemanth this looks good! Although 2 things:

- Remove `print` that was used for debugging.
- Add test that shows that it's doing the right thing https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/POTUS.t

**Comment by hemanth** [comment]
@jagtalon This is still under progress (made this PR for a quick review), need help on analysis why this goodie doesn't get trigger with the current RE.

**Comment by mintsoft** [comment]
@hemanth It doesn't pass the tests because it's using `handle remainder`, that means that `$_` has the actual trigger removed from the query in the handle (https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1913/files#diff-66449435cf1d3b4e87ded23d450483eaR10). 

If you add `warn($_)` before the regex match then you should see what `$_` is

**Comment by hemanth** [comment]
Did try that.

```pl
warn($_);
my $rem = shift;
warn($rem)
```
^ isn't logging anything, I just notice: 

```
"Sorry, no hit for your instant answer"                                                                                                                                         
"GET /?q=potus HTTP/1.1"
```

Am I missing the obvious?

**Comment by mintsoft** [comment]
@hemanth no that's exactly showing the problem. For the query `potus` the remainder isempty, so nothing is output.

Try:

```perl
warn("The Remainder is: $_ \n");
my $rem = shift;
warn("The rem is: $rem \n")
```
and you'll see something like:
```
Query: potus
  You entered: potus
The remainder is:
The rem is:
Sorry, no hit on your instant answer
```

**Comment by hemanth** [comment]
It wasn't just empty string, I also had an error:  `masks earlier declaration in same scope` after fixing that I noticed empty string.

**Comment by mintsoft** [comment]
@hemanth yeah

**Comment by hemanth** [comment]
@mintsoft Yet to work on the test case, but I feel the RE part still needs a fix, it does not trigger for few of the quires.

File: lib/DDG/Goodie/POTUS.pm [hunk 2]
```diff
@@ -25,19 +25,13 @@ attribution github  => ['numbertheory', 'John-Peter Etcheber'],
...
+handle query_clean => sub {
     my $rem = shift;
-    $rem =~ s/
-      |who\s+(is|was)\s+the\s+
-      |^POTUS\s+
-      |\s+(POTUS|president\s+of\s+the\s+united\s+states)$
-      |^(POTUS|president\s+of\s+the\s+united\s+states)\s+
-    //gix;
-
+    return unless my ($tense, $index, $text) = $rem =~ /^(?:who (?:is|was) the )?(\d+(?:st|nd|rd|th))?(?:potus|president of the (?:us|united states|usa)) ?(\d+)?$/;
```
Comment (nilnilnil) [inline lib/DDG/Goodie/POTUS.pm]: It doesn't look like `$index` is used here. It's also not very clear whether or not the ordinal suffix is required for that part of the RE to work. It's also confusing in the context of the final `\d` capture.

I think a comment outlining exactly how this works, and test cases for the permutations, would be a sufficient way of solving that issue.

Thanks for doing this.

Comment (hemanth) [inline lib/DDG/Goodie/POTUS.pm]: Thanks for the review, with the current RE not all the legitimate quires trigger the IA.

**Comment by mintsoft** [comment]
Ping @hemanth Just checking in; are you interested in finishing this ofF?

File: t/POTUS.t [hunk 3]
```diff
@@ -90,6 +90,9 @@ ddg_goodie_test(
             result    => 'Abraham Lincoln'
         }
     ),
+    'vice president of the united states' => test_zci(
+        ''
```
Comment (GuiltyDolphin) [inline t/POTUS.t]: Why is this blank? If it shouldn't trigger we should be using `'...' => undef,`

**Comment by hemanth** [comment]
@mintsoft The last time I tried this, there was a triggering issue.

**Comment by GuiltyDolphin** [comment]
@hemanth Can you expand? Maybe we can help you with that? If you need any Regex help myself and @mintsoft are fairly well versed.

**Comment by hemanth** [comment]
@GuiltyDolphin I ran the server and the I.A wasn't getting triggered.

**Comment by GuiltyDolphin** [comment]
@hemanth Was it triggering on `duckpan query`? It looks like the tests have failed so it is probably the Goodie logic.

**Comment by hemanth** [comment]
@GuiltyDolphin : 

```sh
Query: potus                                                                                                                                                                                              
  You entered: potus                                                                                                                                                                                      
---                                                                                                                                                                                                       
DDG::ZeroClickInfo  {                                                                                                                                                                                     
    Parents       WWW::DuckDuckGo::ZeroClickInfo                                                                                                                                                          
    public methods (3) : has_structured_answer, new, structured_answer                                                                                                                                    
    private methods (0)                                                                                                                                                                                   
    internals: {                                                                                                                                                                                          
        answer              "Barack Obama is the 44th President of the United States.",                                                                                                                   
        answer_type         "potus",                                                                                                                                                                      
        caller              "DDG::Goodie::POTUS",                                                                                                                                                         
        is_cached           1,                                                                                                                                                                            
        is_unsafe           0,                                                                                                                                                                            
        structured_answer   {                                                                                                                                                                             
            input       [                                                                                                                                                                                 
                [0] "44th"                                                                                                                                                                                
            ],                                                                                                                                                                                            
            operation   "President of the United States",                                                                                                                                                 
            result      "Barack Obama"                                                                                                                                                                    
        }                                                                                                                                                                                                 
    }                                                                                                                                                                                                     
}                                                                                                                                                                                                         
---                                                                                                                                                                                                       
Query: potus 16                                                                                                                                                                                           
  You entered: potus 16                                                                                                                                                                                   
---                                                                                                                                                                                                       
DDG::ZeroClickInfo  {                                                                                                                                                                                     
    Parents       WWW::DuckDuckGo::ZeroClickInfo                                                                                                                                                          
    public methods (3) : has_structured_answer, new, structured_answer                                                                                                                                    
    private methods (0)                                                                                                                                                                                   
    internals: {                                                                                                                                                                                          
        answer              "Barack Obama is the 44th President of the United States.",                                                                                                                   
        answer_type         "potus",                                                                                                                                                                      
        caller              "DDG::Goodie::POTUS",                                                                                                                                                         
        is_cached           1,                                                                                                                                                                            
        is_unsafe           0,                                                                                                                                                                            
        structured_answer   {                                                                                                                                                                             
            input       [                                                                                                                                                                                 
                [0] "44th"                                                                                                                                                                                
            ],                                                                                                                                                                                            
            operation   "President of the United States",                                                                                                                                                 
            result      "Barack Obama"                                                                                                                                                                    
        }                                                                                                                                                                                                 
    }                                                                                                                                                                                                     
}                                                                                                                                                                                                         
---                                                                                                                                                                                                       
Query:                                                                                                                                                                                                    
```

File: lib/DDG/Goodie/POTUS.pm [hunk 4]
```diff
@@ -25,19 +25,13 @@ attribution github  => ['numbertheory', 'John-Peter Etcheber'],
...
     my $num = ($rem =~ /^\d+$/) ? $rem : words2nums($rem) || $prez_count;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/POTUS.pm]: `$rem` will have the full text as we are using `query_clean`, so it will always resort to the current count.

Comment (hemanth) [inline lib/DDG/Goodie/POTUS.pm]: What do you suggest here?

File: lib/DDG/Goodie/POTUS.pm [hunk 5]
```diff
@@ -25,19 +25,13 @@ attribution github  => ['numbertheory', 'John-Peter Etcheber'],
...
 
-handle remainder => sub {
+handle query_clean => sub {
     my $rem = shift;
-    $rem =~ s/
-      |who\s+(is|was)\s+the\s+
-      |^POTUS\s+
-      |\s+(POTUS|president\s+of\s+the\s+united\s+states)$
-      |^(POTUS|president\s+of\s+the\s+united\s+states)\s+
-    //gix;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/POTUS.pm]: @hemanth Originally this was 'cleaning' the query to get rid of the unwanted parts and only leave the number. I think we want to reintroduce this.

Comment (hemanth) [inline lib/DDG/Goodie/POTUS.pm]: @GuiltyDolphin But the new RE must cover this and should not trigger for `vice president of the united states`

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/POTUS.pm]: @hemanth The new re doesn't perform any substitutions. Are we sure we want to be grabbing the number from `$rem`? We could try using `$index` instead.

Comment (hemanth) [inline lib/DDG/Goodie/POTUS.pm]: `index` should be fine as well, [this](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1913/files/41d57b1e007a0c1b45e40dd5aa73f03c51a02a27#diff-66449435cf1d3b4e87ded23d450483eaR30) doesn't look ok?

**Comment by mintsoft** [comment]
@GuiltyDolphin I've assigned you to this one; I'm assigned to the other POTUS PR: #2184 Feel free to close #2184 if this is resolved in here

**Comment by jdorweiler** [comment]
ping @GuiltyDolphin

**Comment by GuiltyDolphin** [comment]
Fixed in #2925.

**Pull request closed without merging**